binary_gentoo/internal/cli/build.py

```diff
@@ -192,7 +229,10 @@ def build(config):
     ]
 
     # Create pretend log dir
-    category, package = extract_category_package_from(config.atom)
+    try:
+        category, package = extract_category_package_from(config.atom)
+    except ValueError:
+        category, package = 'set', extract_category_package_from(config.atom)
```

Comment:
Maybe we need a third field with some enum whether we have a set or true package. What feels most wrong about this is the de-facto but unintended support for `set/world/flavors.yml` as a side effect.

Reply:
What would you think about the following?

```
class PackageOrSetEnum:
    package = 1
    set = 2

[.....]

try:
    category, package_or_set = extract_category_package_from(config.atom)
    atom_type = PackageOrSetEnum.package
except ValueError:
    try:
        category, package_or_set = 'set', extract_set_from(config.atom)
        atom_type = PackageOrSetEnum.set
    except ValueError:
        raise ValueError(f'Not valid syntax for package ("{ATOM_LIKE_DISPLAY}") '
                         f'or set ("{SET_DISPLAY}"): {config.atom!r}')
```
and then based on `atom_type` the creation of `flavors.yml` can be executed only for packages and not for sets.
This also renames `package` to `package_or_set` to make it more explicit this could be either. It keeps however the default value of `'set'` for `category` if the atom is a set.
Not sure though about the terminology; if `atom` (for `atom_type`) is a good name for something that can both be category/package or set?

Reply:
Hi!

This will not be short, but it seems needed:


# About a common term for both atoms and sets

In `man emerge` I noticed this part near the very top in section `SYNOPSIS`:

> `emerge [options] [action] [ebuild | tbz2file | file | @set | atom] ...`

So as far as portage is concerned, a set is not a kind of atom but a different thing.
For a term, that includes both, I guess our options include:

- "object"
- "task"
- "job"
- "request"
- "target"
- "item"
- "thing"

Most of these terms have some issue or ambiguity.  Which of these feel least wrong to you?  Did I miss any?


# About atoms in binary-gentoo

What binary-gentoo callas atoms are atom-like but not full atoms, meaning they support

- `cat/pkg` (true atom)
- `cat/pgk-123` (not a true atom)
- `=cat/pgk-123` (true atom)

but not anything like

- `>=cat/pkg-123`
- `cat/pkg-123:0/456`
- `pkg`

My impression is that terms "cp" and "cpv" are mostly used internally in implemenations,
but if we wanted to be more precise we could pass `metavar='CP|CPV|=CPV|@SET'`
to `argparse` and call it `atom_or_set` in `dest` internally.


# Category "set"

Maybe it's okay to be practical rather than purist there.  Would you be okay with plural?
Gentoo package categories like `dev-util` seem to be singular but I see e.g. `/etc/portage/sets/` in `man portage`.


# Enums

We have an enum with two values here.  Options that I consider natural are:

- a) a single bool e.g. `is_package = True`
- b) two uppercase "constants", e.g. `TYPE_PACKAGE, TYPE_SET = range(2)`
- c) a [true Python 3 enum](https://docs.python.org/3/library/enum.html#creating-an-enum)
- d) …

(a) upside of a is simplicity, downside is bad extensibility and that "not is_package" doesn't say what the opposite is.  What do you think about these three?


# Use of variable "category"

I checked where `category = "set"` would have effects and I cam to this data flow tree:

- `category`
  - `host_pretend_logdir_category`
    - `host_pretend_logdir_category_package`
      - `pretend_fail_log_filename`
  - `flavors_yml_file`

As long as we make sure sets do not go to `flavors_yml_file`, the rest seems fine.
Please note that the "pretend" in the variable names is an out-of-sync leftover
that I should remove.  Let me do a PR on purely that in a minute.


# Exception messages

It would be nice to have exception messages be an uppercase-starting sentence
and to include the bad value in the message.  If that exceptions ends up at the user terminal, that will go a long way.

Reply:
How do you feel about calling the argument `emerge_target` instead of `atom` ? It's a little more specific than just `target` for someone who reads the code, but does not try to say what it exactly is (atom, atomlike, set, ...). 
About the enums: personally I feel that using an enum instead of a boolean for this kind of situation gives more flexibility and room for changes later on. So I'd prefer to go with option (c) to make a true Python3 enum.

Reply:
Good idea about `emerge_target`.
Sounds like Python3 enums it is.
Cool!